packages/contract/src/pay/DaimoPay.sol

```diff
@@ -0,0 +1,269 @@
+        address indexed intentAddr,
+        address indexed newRecipient,
+        PayIntent intent
+    );
+
+    // Action settled later, once the underlying bridge transfer completes.
+    event Claim(
+        address indexed intentAddr,
+        address indexed finalRecipient,
+        PayIntent intent
+    );
+
+    // When the action is completed as expected, emit this event
+    event ActionCompleted(
+        address indexed intentAddr,
+        address indexed destinationAddress
+    );
+
+    // When the action is a call that fails, we bounce the funds to the
+    // specified refund address and emit this event
+    event ActionBounced(
+        address indexed intentAddr,
+        address indexed refundAddress
+    );
+
+    constructor(PayIntentFactory _intentFactory, DaimoPayBridger _bridger) {
+        intentFactory = _intentFactory;
+        bridger = _bridger;
+    }
+
+    // Helper functions to convert between address and bytes32
+    // Solana and Tron addresses are 32 bytes, but Ethereum addresses are 20 bytes.
+    function addressToBytes32(address addr) internal pure returns (bytes32) {
+        return bytes32(uint256(uint160(addr)));
+    }
+
+    function bytes32ToAddress(bytes32 b) internal pure returns (address) {
+        return address(uint160(uint256(b)));
+    }
+
+    function startAction(
+        PayIntent calldata intent,
+        Call[] calldata calls,
+        bytes calldata bridgeExtraData
+    ) public {
+        PayIntentContract intentContract = intentFactory.createIntent(intent);
+
+        // Ensure we don't reuse a nonce in the case where Alice is sending to
+        // same destination with the same nonce multiple times.
+        require(!intentSent[address(intentContract)], "DP: already sent");
+        intentSent[address(intentContract)] = true;
+
+        intentContract.sendAndSelfDestruct(
+            intent,
+            bridger,
+            calls,
+            bridgeExtraData
+        );
+
+        emit Start({intentAddr: address(intentContract), intent: intent});
+    }
+
+    // Pays Bob immediately on chain B. The caller LP sends (toToken, toAmount).
+    // Later, when the slower bridge transfer arrives, the LP will be able to claim
+    // (toToken, fromAmount), keeping the spread (if any) between the amounts.
+    function fastFinishAction(
+        PayIntent calldata intent,
+        Call[] calldata calls
+    ) public {
+        require(intent.toChainId == block.chainid, "DP: wrong chain");
+
+        // Calculate handoff address
+        address intentAddr = intentFactory.getIntentAddress(intent);
+
+        // Optimistic fast finish is only for transfers which haven't already
+        // been fastFinished or claimed.
+        require(
+            intentToRecipient[intentAddr] == address(0),
+            "DP: already finished"
+        );
+
+        // Record LP as new recipient
+        intentToRecipient[intentAddr] = msg.sender;
+
+        // LP fast-deposits bridgeTokenOut
```

Comment:
does LP always send bridgeTokenOut vs the finalToken ?

Reply:
one more idea: we remove this part and go directly to `completeAction`
- if LP wants to pay in bridgeTokenOut + swap, they transfer it themselves before calling `fastFinish..`
- alternately, if LP wants to pay in `finalCallToken` directly, they can do that too + pass an empty `calls` array
- either way, completeAction will ensure that enough of the `finalCallToken` is present